File: app/src/main/java/com/mgaetan89/showsrage/fragment/ShowFragment.java
package com.mgaetan89.showsrage.fragment;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.mgaetan89.showsrage.Constants;
import com.mgaetan89.showsrage.R;
import com.mgaetan89.showsrage.adapter.ShowPagerAdapter;
import com.mgaetan89.showsrage.model.Seasons;
import com.mgaetan89.showsrage.model.Show;
import com.mgaetan89.showsrage.network.SickRageApi;

import java.util.ArrayList;
import java.util.List;

import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

public class ShowFragment extends Fragment implements Callback<Seasons> {
	@Nullable
	private ShowPagerAdapter adapter = null;

	@NonNull
	private final List<Integer> seasons = new ArrayList<>();

	@Nullable
	private TabLayout tabLayout = null;

	@Nullable
	private ViewPager viewPager = null;

	public ShowFragment() {
	}

	@Override
	public void failure(RetrofitError error) {
		error.printStackTrace();
	}

	@Override
	public void onActivityCreated(@Nullable Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);

		Intent intent = this.getActivity().getIntent();
		Show show = intent.getParcelableExtra(Constants.Bundle.SHOW_MODEL);
		String sort = getSeasonsSort(PreferenceManager.getDefaultSharedPreferences(this.getContext()));

		SickRageApi.getInstance().getServices().getSeasons(show.getIndexerId(), sort, this);

		this.tabLayout = (TabLayout) this.getActivity().findViewById(R.id.tabs);

		if (this.tabLayout != null && this.viewPager != null) {
			this.tabLayout.setupWithViewPager(this.viewPager);
			this.tabLayout.setVisibility(View.VISIBLE);
		}
	}

	@Nullable
	@Override
	public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_show, container, false);

		if (view != null) {
			this.viewPager = (ViewPager) view.findViewById(R.id.show_pager);

			if (this.viewPager != null) {
				this.adapter = new ShowPagerAdapter(this.getChildFragmentManager(), this, this.seasons);

				this.viewPager.setAdapter(this.adapter);
			}
		}

		return view;
	}

	@Override
	public void onDestroy() {
		this.seasons.clear();

		super.onDestroy();
	}

	@Override
	public void onDestroyView() {
		this.tabLayout = null;
		this.viewPager = null;

		super.onDestroyView();
	}

	@Override
	public void success(Seasons seasons, Response response) {
		this.seasons.clear();
		this.seasons.addAll(seasons.getData());

		if (this.adapter != null) {
			this.adapter.notifyDataSetChanged();

			if (this.tabLayout != null) {
				this.tabLayout.setTabsFromPagerAdapter(this.adapter);
			}
		}
	}

	/* package */
	static String getSeasonsSort(@Nullable SharedPreferences preferences) {
		if (preferences == null) {
			return "desc";
		}

		return preferences.getBoolean("display_seasons_sort", false) ? "asc" : "desc";
	}
}

File: app/src/main/kotlin/com/mgaetan89/showsrage/fragment/ShowFragment.kt
package com.mgaetan89.showsrage.fragment

import android.os.Bundle
import android.support.v4.view.PagerAdapter
import com.mgaetan89.showsrage.Constants
import com.mgaetan89.showsrage.R
import com.mgaetan89.showsrage.activity.MainActivity
import com.mgaetan89.showsrage.adapter.ShowPagerAdapter
import com.mgaetan89.showsrage.extension.getPreferences
import com.mgaetan89.showsrage.extension.getSeasonSort
import com.mgaetan89.showsrage.extension.getShow
import com.mgaetan89.showsrage.model.Seasons
import com.mgaetan89.showsrage.model.Sort
import com.mgaetan89.showsrage.network.SickRageApi
import io.realm.Realm
import retrofit.Callback
import retrofit.RetrofitError
import retrofit.client.Response

class ShowFragment : TabbedFragment(), Callback<Seasons> {
    private val seasons = mutableListOf<Int>()

    override fun failure(error: RetrofitError?) {
        error?.printStackTrace()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        val activity = this.activity

        if (activity is MainActivity) {
            activity.displayHomeAsUp(true)
            activity.setTitle(R.string.show)
        }

        val indexerId = this.arguments.getInt(Constants.Bundle.INDEXER_ID)
        val realm = Realm.getDefaultInstance()
        val show = realm.getShow(indexerId)
        val sort = activity.getPreferences().getSeasonSort()
        val seasons = show?.seasonList?.map { it.value.toInt() } ?: emptyList()

        realm.close()

        this.displaySeasons(if (Sort.ASCENDING.equals(sort)) seasons.sorted() else seasons.sortedDescending())

        SickRageApi.instance.services?.getSeasons(indexerId, sort.label, this)
    }

    override fun onDestroy() {
        this.seasons.clear()

        super.onDestroy()
    }

    override fun success(seasons: Seasons?, response: Response?) {
        this.displaySeasons(seasons?.data)
    }

    override fun getAdapter(): PagerAdapter {
        return ShowPagerAdapter(this.childFragmentManager, this, this.seasons)
    }

    override fun useSwipeToRefresh() = false

    private fun displaySeasons(seasons: Iterable<Int>?) {
        this.seasons.clear()
        this.seasons.addAll(seasons ?: emptyList())

        this.updateState(this.seasons.isEmpty())
    }
}
